src/interrupt/mod.rs

```diff
@@ -0,0 +1,188 @@
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(e: io::Error) -> Error {
+        Error::IOError(e)
+    }
+}
+
+/// Trait for interrupt producers.
+pub trait Interrupt {
+    /// The type of Trigger object used for notifications by this interrupt.
+    type TriggerType: Trigger;
+
+    /// Inject an interrupt from this interrupt source into the guest.
+    fn trigger(&self) -> std::result::Result<(), <Self::TriggerType as Trigger>::E>;
+
+    /// Returns an interrupt notifier from this interrupt.
+    ///
+    /// An interrupt notifier allows for external components and processes
+    /// to inject interrupts into a guest, by writing to the file returned
+    /// by this method.
+    fn notifier(&self) -> Option<Self::TriggerType> {
+        None
+    }
+
+    /// Called back when the CPU acknowledges the interrupt.
+    fn acknowledge(&self) {}
+
+    /// Returns an end-of-interrupt notifier from this interrupt.
+    ///
+    /// An end-of-interrupt notifier allows for external components and processes
+    /// to be notified when a guest acknowledges an interrupt.  This can be used
+    /// to resample and inject a level-triggered interrupt, or to mitigate the
+    /// effect of lost timer interrupts.
+    fn ack_notifier(&self) -> Option<Self::TriggerType> {
+        None
+    }
+
+    /// Enable generation of interrupts from this interrupt source.
+    fn enable(&self) -> Result<()>;
+
+    /// Disable generation of interrupts from this interrupt source.
+    fn disable(&self) -> Result<()> {
+        Err(Error::OperationNotSupported)
+    }
+}
+
+/// Trait for interrupts that allow users to configure interrupt parameters.
+pub trait ConfigurableInterrupt: Interrupt {
+    /// Type describing the configuration spec of the interrupt.
+    type Cfg;
+
+    /// Update configuration of the interrupt.
+    fn update(&self, config: &Self::Cfg) -> Result<()>;
+
+    /// Returns the current configuration of the interrupt.
+    fn get_config(&self) -> Result<Self::Cfg>;
+}
+
+/// Trait for interrupts that can be masked or unmasked.
+pub trait MaskableInterrupt: Interrupt {
+    /// Mask the interrupt.  Masked interrupts are remembered but
+    /// not delivered.
+    fn mask(&self) -> Result<()>;
+
+    /// Unmask the interrupt, delivering it if it was pending.
+    fn unmask(&self) -> Result<()>;
+}
+
+/// Trait to manage a group of interrupt sources for a device.
+///
+/// A device may use an InterruptSourceGroup to manage multiple interrupts of the same type.
+/// The group allows a device to request and release interrupts and perform actions on the
+/// whole collection of interrupts like enable and disable for cases where enabling or disabling
+/// a single interrupt in the group does not make sense. For example, PCI MSI interrupts must be
+/// enabled as a group.
+pub trait InterruptSourceGroup: Send {
+    /// Type of the interrupts contained in this group.
+    type InterruptType: Interrupt;
```

Comment:
Is there a real need for this associated type? Couldn't we get away with only `InterruptWrapper`?

Reply:
I think we need it so that we can have a bound for the Deref target.